dsa-solutions/lc-solutions/0000-0099/0072-edit-distance.md

```diff
@@ -0,0 +1,166 @@
+exection -> execution (insert 'u')
+```
+
+### Constraints
+
+- `0 <= word1.length, word2.length <= 500`
+- `word1` and `word2` consist of lowercase English letters.
+
+## Solution
+
+We explore two main approaches: Recursive Dynamic Programming with 
+Memoization and Iterative Dynamic Programming with Tabulation.
+
+### Approach 1: Recursive Dynamic Programming (Memoization)
+Concept: Store the solutions for each position to avoid redundant calculations.
+
+#### Algorithm
+
+For every index of string S1, we have three options to match that index with string S2, i.e replace the character, remove the character or insert some character at that index. Therefore, we can think in terms of string matching path as we have done already in previous questions.
+
+As there is no uniformity in data, there is no other way to find out than to try out all possible ways. To do so we will need to use recursion.
+
+Steps to memoize a recursize solution:
+- Create a dp array of size [n][m]. The size of S1 and S2 are n and m respectively, so the variable i will always lie between ‘0’ and ‘n-1’ and the variable j between ‘0’ and ‘m-1’.
+-  We initialize the dp array to -1.
+Whenever we want to find the answer to particular parameters (say f(i,j)), we first check whether the answer is already calculated using the dp array(i.e dp[i][j]!= -1 ). If yes, simply return the value from the dp array.
+- If not, then we are finding the answer for the given value for the first time, we will use the recursive relation as usual but before returning from the function, we will set dp[i][j] to the solution we get.
+
+#### Implementation
+
+```C++
+int editDistanceUtil(string& S1, string& S2, int i, int j, vector<vector<int>>& dp) {
+    if (i < 0)
+        return j + 1;
+    if (j < 0)
+        return i + 1;
+
+    if (dp[i][j] != -1)
+        return dp[i][j];
+
+    if (S1[i] == S2[j])
+        return dp[i][j] = 0 + editDistanceUtil(S1, S2, i - 1, j - 1, dp);
+
+    else
+        return dp[i][j] = 1 + min(editDistanceUtil(S1, S2, i - 1, j - 1, dp),
+                                  min(editDistanceUtil(S1, S2, i - 1, j, dp),
+                                      editDistanceUtil(S1, S2, i, j - 1, dp)));
+}
+
+int editDistance(string& S1, string& S2) {
+    int n = S1.size();
+    int m = S2.size();
+
+    vector<vector<int>> dp(n, vector<int>(m, -1));
+
+    return editDistanceUtil(S1, S2, n - 1, m - 1, dp);
+}
+```
+
+### Complexity Analysis
+
+- **Time complexity**: O(N*M)
+Reason: There are N*M states therefore at max ‘N*M’ new problems will be solved.
+- **Space complexity**: O(N*M) + O(N+M)
+Reason: We are using a recursion stack space(O(N+M)) and a 2D array ( O(N*M)).
+
+### Approach 2: Iterative Dynamic Programming (Tabulation)
+
+Concept: In the recursive logic, we set the base case too if(i<0 ) and if(j<0) but we can’t set the dp array’s index to -1. Therefore a hack for this issue is to shift every index by 1 towards the right.
```

Comment:
replace `if(i<0 )` to 

```md
`if(i<0 )`
```

and solve error

Reply:
@Ajay-Dhangar updated the code as per your suggestion. Kindly review PR again.